---
title: "Untitled"
author: "Benji Lamp"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE, echo=F}
knitr::opts_chunk$set(echo = TRUE)

library(DESeq2)
library(dplyr)
library(tidyr)
library(kableExtra)
library(here)
library(AnnotationDbi)
library(org.Bt.eg.db)
```

```{r}
input <- "mRNA_deseq_results.RData"
load(here("processed_data", input))
```


# Timecourse Analysis

```{r, echo=FALSE, message=FALSE}
dds <- estimateSizeFactors(dds)
normalized_counts_df <- as.data.frame(counts(dds, normalized=TRUE))
col_names <- rep(c("v", "mp", "lp", "el", "l"), each = 7)
names(normalized_counts_df) <- col_names

# Prepare an empty data frame for storing averages, with correct dimensions and column names
averages_df <- data.frame(matrix(ncol = length(unique(col_names)), nrow = nrow(normalized_counts_df)))
names(averages_df) <- unique(col_names)
rownames(averages_df) <- rownames(normalized_counts_df)

# Loop through each time-point to calculate and store row means
for (timepoint in unique(col_names)) {
  # Identify columns belonging to the current time-point
  cols <- which(col_names == timepoint)
  # Calculate and store
  averages_df[[timepoint]] <- rowMeans(normalized_counts_df[, cols], na.rm = TRUE)
}

```

```{r, echo=FALSE, message=FALSE, results='asis'}
hadf <- head(averages_df)
# Custom CSS to enlarge table captions
cat("
<style>
table caption {
  font-size: 16pt;
  font-weight: bold;
}
</style>
")

# Assuming 'samples' is another data frame you want to display
kable(hadf, caption = "Mean of Normalized Counts") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = FALSE)
```

## Pairwise Comparison

To perform differential expression analysis, comparative structures need to be formed between timepoints. The loop below iterates through each successive comparison and generates 'res' objects from DESeq2, which contain statistics like log fold change and adjusted p-values for each comparison. After the False Discovery Rate adjusted p-value (padj) is generated, the mean of the normalized counts is evaluated between each comparison. If gene expression increases from time n to n+1, an 'I' will be assigned to the gene. If gene expression decreases from n to n+1, a 'D' will be assigned to the gene.

The adjusted p-value generated from DESeq2 is then used to evaluate whether there is significance. If the padj value is greater than 0.05, the gene will be deemed to have an insignificant change in expression between timepoints and will be assigned an 'S'



```{r}
contrasts <- list(
  c("condition", "l", "v"),
  #
  c("condition", "el", "v"),
  c("condition", "l", "mp"),
  #
  c("condition", "lp", "v"),
  c("condition", "el", "mp"),
  c("condition", "l", "lp"),
  #
  c("condition", "mp", "v"),
  c("condition", "lp", "mp"),
  c("condition", "el", "lp"),
  c("condition", "l", "el")
)

contrast_results <- list()

# DE for pairs in contrasts
for (i in seq_along(contrasts)) {
  contrast = contrasts[[i]]
  res <- results(dds,
                 contrast=contrast)
  
  # Determine if the mean expression increases or decreases
  condition1_mean <- averages_df[[contrast[3]]]  # Mean for the first condition
  condition2_mean <- averages_df[[contrast[2]]]  # Mean for the second condition
  expressionDirection <- ifelse(condition1_mean < condition2_mean, "I", "D")
  
  # Hypothesis test, FTR null assign 'S', Rej null assign I/D
  res$expressionChange <- ifelse(res$padj > 0.05, "S", expressionDirection)
  
  contrast_results[[paste(contrast[3], "to", contrast[2])]] <- res
}
```

## Motif Structuring
The variables (I,S,D) that corespond to expression are currently being stored in the list of DESeq objects. In order to create an index of gene expression, genes are added to a dataframe that completley describes their patterns of expression over the time course. This tabe can be manipulated in order to generate a singular index of pattern structure depending on the needs of the research. 


The complete time course analysis can be organized into a piecewise hierarchical structure with 4 tiers, 

- Primary:		V-PL\
- Secondary:	V-EL MP-PL\
- Tertiary:		V-LP MP-EL LP-PL\
- Quaternary:	V-MP MP-LP LP-EL EL-PL\

A complete model vector will be structured as\
Primary-Secondary-Tertiary-Quaternary\
X-XX-XXX-XXXX\


```{r}
combined_results <- data.frame(gene=rownames(contrast_results[[1]]))
# Loop through each contrast result to combine them
for (i in seq_along(contrast_results)) {
  contrast_name <- names(contrast_results)[i]  # Get the name of the current contrast
  # Create a temporary dataframe with genes and their expression changes for the current contrast
  temp_df <- data.frame(gene=rownames(contrast_results[[i]]), 
                        expressionChange=contrast_results[[i]]$expressionChange)
  colnames(temp_df)[2] <- contrast_name  # Rename the second column to the current contrast name
  # Merge the temporary dataframe with the combined_results dataframe
  combined_results <- merge(combined_results, temp_df, by="gene", all=TRUE)
}

# Concatenate expression patterns to make model vector
combined_results$modelVector <- apply(combined_results[, -1], 1, 
                                      function(x) paste(x, collapse = ""))

# for Quaternary tree structure ###############################################################
combined_results$primary <- combined_results[,2]
combined_results$secondary <- paste0(combined_results[,3], combined_results[,4])
combined_results$tertiary <- paste0(combined_results[,5], combined_results[,6], combined_results[,7])
combined_results$quaternary <- paste0(combined_results[,8], combined_results[,9], combined_results[,10], combined_results[,11])



```

```{r, echo=FALSE, message=FALSE, results='asis'}
relevant_data <- combined_results %>%
  dplyr::select(starts_with("v to"), starts_with("mp to"), starts_with("lp to"), starts_with("el to"))

# Reshape data from wide to long format
long_data <- relevant_data %>%
  pivot_longer(cols = everything(), names_to = "transition", values_to = "event") %>%
  group_by(transition, event) %>%
  summarise(count = n(), .groups = 'drop')

# Spread the data back to wide format to see counts for each event type in each transition
wide_data <- long_data %>%
  pivot_wider(names_from = event, values_from = count, values_fill = list(count = 0))


wide <- as.data.frame(wide_data)

cat("
<style>
table caption {
  font-size: 8pt;
}
</style>
")

# Assuming 'samples' is another data frame you want to display
kable(wide, caption = "Incedence of Increasing, Decreasing and Steady Genes") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = FALSE)
```

```{r, echo=FALSE, message=FALSE, results='asis'}
selected_columns <- combined_results[, c(1, 12:16)]

# Custom CSS to enlarge table captions
cat("
<style>
table caption {
  font-size: 8pt;
  font-weight: bold;
}
</style>
")

# Assuming 'samples' is another data frame you want to display
kable(head(combined_results[,1:11]), caption = "") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = FALSE)

kable(head(selected_columns), caption = "First 5 Rows of Pattern Dataframe") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = FALSE)


```

## Mapping Specific Motif

Reducing the number of comparisons within a pattern structure can significantly decrease the total number of patterns identified. Additionally, depending on the objectives of the experiment, different patterns can be constructed that reveal more biologically relevant structures. Below, the 'vector' variable was constructed to create patterns from virgin to pregnancy to lactation.

```{r}
motif_index <- list()
x <- combined_results
vector <- paste0(x$`v to mp`,
                 x$`v to lp`,
                 x$`mp to el`,
                 x$`lp to el`,
                 x$`mp to l`,
                 x$`lp to l`,
                 x$`el to l`)

# Loop through each unique modelVector
for(model in unique(vector)) {
  # Subset the genes that match the current modelVector
  genes <- combined_results$gene[vector == model]
  # Store
  motif_index[[model]] <- genes
}

# Specific patterns can be called using this function 
head(motif_index$SIIIIIS)
```

Throughout this entire process, Ensembl IDs have been used as gene identifiers. To make the data easier to read, mapping to symbols can be done using AnnotationDbi and the org.Bt.eg.db package for bovine gene symbols.

```{r, message=F, warning=F}
mapped_motif_index <- list()
for(model in names(motif_index)) {
  ensembl_ids <- motif_index[[model]]
  gene_symbols <- tryCatch({
    mapIds(org.Bt.eg.db,
           keys = ensembl_ids,
           column = "SYMBOL",
           keytype = "ENSEMBL",
           multivals = 'first')
  }, error = function(e) {
    message("Error with model ", model, ": ", e$message)
    return(NA)  # Return NA or any other placeholder that indicates an error
  })
  mapped_motif_index[[model]] <- gene_symbols
}
head(mapped_motif_index$SIIIIIS)

```
